### PR TITLE
Shield Activity

### PR DESCRIPTION
**Pull request by panleone**
## Abstract

Implements shield activity, i.e. shield transactions regarding the wallet are decrypted and loaded in the dashboard activity.

 3 new transactions types have been added:

- Shielding, which is a self (t -> s) transaction;
- De-Shielding, which is a self (s->t) transaction;
- Shield sent to self, which is a self (s->s) transaction.

shield_activity

  Transactions received and sent to external wallets are marked as before, with the addition of shield addresses:
- Sent to (transparent or shield address);
- Received with (transparent or shield address)

I suggest to review the PR commit by commit, is easier than looking at the overall diff. Also to test the PR you have to manually build and use the following version of the pivx-shield library: [PR 81](https://github.com/PIVX-Labs/pivx-shield/pull/81).

---

## Testing
To test this PR, it's suggested to attempt these user flows, or variations of these:
- Create and sync a shield wallet;
- Create the 3 new types of transaction;
- Send/Receive shielded pivs from an external wallet;
- Close and re-open the wallet: the shield activity should be unchanged;
- Resync the wallet from zero: the shield activity should be unchanged.

### CODE REVIEW

**Comment by Duddino** [comment]
```diff
diff --git a/package.json b/package.json
index 547b11a..4b044ca 100644
--- a/package.json
+++ b/package.json
@@ -89,9 +89,9 @@
         "jquery": "^3.6.3",
         "pinia": "^2.1.7",
         "pivx-promos": "^0.2.0",
-        "pivx-shield": "^1.1.6",
-        "pivx-shield-rust": "^1.1.6",
-        "pivx-shield-rust-multicore": "^1.1.10",
+        "pivx-shield": "^1.2.0-0",
+        "pivx-shield-rust": "^1.2.0-0",
+        "pivx-shield-rust-multicore": "^1.2.0-1",
         "qr-scanner": "^1.4.2",
         "qrcode-generator": "^1.4.4",
         "vue": "^3.3.4",
```
I published a beta version with pivx-shield#81, apply this patch and `npm i` to use

File: scripts/dashboard/Activity.vue [hunk 1]
```diff
@@ -61,6 +61,39 @@ const txMap = computed(() => {
     };
 });
 
+/**
+ * Returns the information that we need to show (icon + label + amount) for a self transaction
+ * @param {number} amount - The net amount of transparent PIVs in a transaction
+ * @param {number} shieldAmount - The net amount of shielded PIVs in a transaction
+ */
+function txSelfMap(amount, shieldAmount) {
```
Comment (Duddino) [inline scripts/dashboard/Activity.vue:69]: Follow the design colors/icons

File: scripts/wallet.js [hunk 2]
```diff
@@ -672,15 +676,54 @@ export class Wallet {
...
+        }
+
+        for (const shieldSpend of tx.shieldSpend) {
+            //TODO: find a better way to reverse and swap endianess
+            const stringArr = shieldSpend.nullifier.split('').reverse();
+            for (let i = 0; i < stringArr.length - 1; i += 2) {
+                const temp = stringArr[i + 1];
+                stringArr[i + 1] = stringArr[i];
+                stringArr[i] = temp;
+            }
```
Comment (Duddino) [inline scripts/wallet.js:710]: You can convert to array, swap that and re-convert to hex

**Review by Duddino — CHANGES_REQUESTED** [review]

File: scripts/wallet.js (hunk 2 above)
Comment (JSKitty) [reply scripts/wallet.js:710]: Imo, we can write an efficient `swapEndianness` util in `util.js`, useful and easier to read to re-use in a compact form than manually swap it for every instance.

**Comment by panleone** [comment]
Reviews applied:

- Moved the reverse and swap endianess to utils.js
- Fixed the bug where some txs were markes as "Sent to shield address", even only the change was shielded
-  Made wallets automatically resync when they use the new shielding version for the first time

**Review by JSKitty — DISMISSED** [review]
**tACK** 9e81bae61091277951d06cb341d3e83675ce825b

Working awesome - two Quality Control members also found no issues, smooth sailing, activity loading is much faster with the new optimisations done since the last Shield Activity tests.

**Review by Duddino — DISMISSED** [review]
tACK

**Review by Duddino — APPROVED** [review]
tACK

**Review by JSKitty — APPROVED** [review]
**tACK** ead84d0e837d7e460b43c53b50d2ebe270801586

**Pull request merged**